src/main/java/gregtech/common/metatileentities/storage/CraftingRecipeLogic.java

```diff
@@ -0,0 +1,198 @@
+            craftingGrid.setStackInSlot(i, ingredients.getOrDefault(i + 1, ItemStack.EMPTY));
+        }
+    }
+
+    private boolean hasCraftingGridUpdated() {
+        boolean craftingGridChanged = false;
+        for (int i = 0; i < craftingGrid.getSlots(); i++) {
+            ItemStackKey oldStack = oldCraftingGrid[i];
+            ItemStack newStack = craftingGrid.getStackInSlot(i);
+            if (oldStack == null) {
+                if (newStack.isEmpty()) {
+                    continue;
+                }
+                oldStack = KeySharedStack.getRegisteredStack(newStack);
+                oldCraftingGrid[i] = oldStack;
+                inventoryCrafting.setInventorySlotContents(i, newStack.copy());
+                craftingGridChanged = true;
+            } else if (newStack.isEmpty()) {
+                oldCraftingGrid[i] = null;
+                inventoryCrafting.setInventorySlotContents(i, ItemStack.EMPTY);
+                craftingGridChanged = true;
+            } else if (!ItemStack.areItemsEqual(oldStack.getItemStackRaw(), newStack) ||
+                    !ItemStack.areItemStackTagsEqual(oldStack.getItemStackRaw(), newStack)) {
+                oldCraftingGrid[i] = KeySharedStack.getRegisteredStack(newStack);
+                inventoryCrafting.setInventorySlotContents(i, newStack.copy());
+                craftingGridChanged = true;
+            }
+        }
+        return craftingGridChanged;
+    }
+
+    public void performRecipe(EntityPlayer player) {
+        if (!isRecipeValid()) {
+            return;
+        }
+        if (!cachedRecipeData.consumeRecipeItems()) {
+            return;
+        }
+        ForgeHooks.setCraftingPlayer(player);
+        NonNullList<ItemStack> remainingItems = cachedRecipe.getRemainingItems(inventoryCrafting); // todo right here is where tools get damaged (in UI)
+        ForgeHooks.setCraftingPlayer(null);
+        for (int i = 0; i < remainingItems.size(); i++) {
+            ItemStack itemStack = remainingItems.get(i);
+            if (itemStack.isEmpty()) {
+                continue;
+            }
+            ItemStackKey stackKey = KeySharedStack.getRegisteredStack(itemStack);
+            inventoryCrafting.setInventorySlotContents(i, itemStack);
+            int remainingAmount = itemStack.getCount() - itemSources.insertItem(stackKey, itemStack.getCount(), false, IItemList.InsertMode.HIGHEST_PRIORITY);
+            if (remainingAmount > 0) {
+                itemStack.setCount(remainingAmount);
+                player.addItemStackToInventory(itemStack);
+                if (itemStack.getCount() > 0) {
+                    player.dropItem(itemStack, false, false);
+                }
+            }
+        }
+    }
+
+    public void handleItemCraft(ItemStack itemStack, EntityPlayer player, boolean simulate) {
+        itemStack.onCrafting(world, player, 1);
+        itemStack.getItem().onCreated(itemStack, world, player);
```

Comment:
Should this be put behind the simulate check so it is called if the item is actually created, or is it fine outside of the check?

Reply:
i actually dont see in the code where simulate is true on this case
if `canTakeStack` returns true `onItemTake` is called with simulate false